include/xtensor-sparse/xmap.hpp

```diff
@@ -0,0 +1,285 @@
+    const typename xmap_container<T>::value_type xmap_container<T>::ZERO = 0;
+
+    template<class T>
+    inline xmap_container<T>::xmap_container(const shape_type& shape)
+     : m_shape{shape}
+    {}
+
+    template <class T>
+    inline auto xmap_container<T>::size() const noexcept -> size_type
+    {
+        return compute_size(shape());
+    }
+
+    template<class T>
+    inline auto xmap_container<T>::shape() const -> const shape_type&
+    {
+        return m_shape;
+    }
+
+    template<class T>
+    inline auto xmap_container<T>::dimension() const -> size_type
+    {
+        return m_shape.size();
+    }
+
+    template <class T>
+    template <class S>
+    inline void xmap_container<T>::resize(S&& shape, bool force)
+    {
+        std::size_t dim = shape.size();
+        if (m_shape.size() != dim || !std::equal(std::begin(shape), std::end(shape), std::begin(m_shape)) || force)
+        {
+            m_shape = xtl::forward_sequence<shape_type, S>(shape);
+        }
+    }
+
+    template <class T>
+    template <class S>
+    inline void xmap_container<T>::resize(S&& shape)
+    {
+        resize(std::forward<S>(shape), true);
+    }
+
+    template <class T>
+    template <class S>
+    inline auto& xmap_container<T>::reshape(S&& shape) &
+    {
+        reshape_impl(std::forward<S>(shape), std::is_signed<std::decay_t<typename std::decay_t<S>::value_type>>());
+        return *this;
+    }
+
+    template <class T>
+    template <class TT>
+    inline auto& xmap_container<T>::reshape(std::initializer_list<TT> shape) &
+    {
+        using sh_type = rebind_container_t<T, shape_type>;
+        sh_type sh = xtl::make_sequence<sh_type>(shape.size());
+        std::copy(shape.begin(), shape.end(), sh.begin());
+        reshape_impl(std::move(sh), std::is_signed<TT>());
+        return *this;
+    }
+    
+    template <class T>
+    template <class S>
+    inline void xmap_container<T>::reshape_impl(S&& shape, std::false_type /* is unsigned */)
+    {
+        if (compute_size(shape) != this->size())
+        {
+            XTENSOR_THROW(std::runtime_error, "Cannot reshape with incorrect number of elements. Do you mean to resize?");
+        }
+
+        update_keys(std::forward<S>(shape));
+        m_shape = xtl::forward_sequence<shape_type, S>(shape);
+    }
+
+    template <class T>
+    template <class S>
+    inline void xmap_container<T>::reshape_impl(S&& _shape, std::true_type /* is signed */)
+    {
+        using value_type = typename std::decay_t<S>::value_type;
+        auto new_size = compute_size(_shape);
+        if (this->size() % new_size)
+        {
+            XTENSOR_THROW(std::runtime_error, "Negative axis size cannot be inferred. Shape mismatch.");
+        }
+        std::decay_t<S> shape = _shape;
+        value_type accumulator = 1;
+        std::size_t neg_idx = 0;
+        std::size_t i = 0;
+        for(auto it = shape.begin(); it != shape.end(); ++it, i++)
+        {
+            auto&& dim = *it;
```

Comment:
I think that `auto` is enough here.

Reply:
this line comes from the implementation of `reshape_impl` for an `xstrided_container`.